File: packages/dart/lib/Seguranca/Ataques/AtaqueAvalancheChecksum.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// O AtaqueColisaoChecksum verifica se há colisões; esse verifica a difusão
/// interna: mudar 1 bit da MENSAGEM autenticada deveria mudar ~50% dos bits
/// dos 32 bytes de MAC, em QUALQUER posição. Um efeito avalanche fraco numa
/// posição específica significa que aquele byte quase não influencia o MAC -
/// uma pista forte de que a mistura tem um "ponto cego" estrutural.
///
/// Varre TODAS as posições/bits da entrada (não amostra aleatória) para
/// apontar exatamente onde está o ponto fraco.
class AtaqueAvalancheChecksum extends Ataque {
  final int _mensagensPorCombinacao;
  final int _tamanhoEntrada;
  final double _limiteMedia;
  final double _limitePiorCaso;

  AtaqueAvalancheChecksum({
    this._mensagensPorCombinacao = 10,
    this._tamanhoEntrada = 32,
    this._limiteMedia = 0.4,
    this._limitePiorCaso = 0.4,
  });

  @override
  String nome() => 'Efeito avalanche do checksum/MAC';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chaveMac = repetir('M', 32);

    final primeiro = alvo.checksumBruto('teste', chaveMac);
    final tamanhoSaida = primeiro.length;

    var somaGlobal = 0.0;
    var combinacoes = 0;
    var piorPosicao = -1;
    var piorBit = -1;
    var piorValor = 1.0;

    for (var pos = 0; pos < _tamanhoEntrada; pos++) {
      for (var bit = 0; bit < 8; bit++) {
        var soma = 0.0;
        for (var m = 0; m < _mensagensPorCombinacao; m++) {
          final entrada = randomBytes(_tamanhoEntrada);
          final alterada = Uint8List.fromList(entrada);
          alterada[pos] = alterada[pos] ^ (1 << bit);

          final h1 = alvo.checksumBruto(entrada, chaveMac);
          final h2 = alvo.checksumBruto(alterada, chaveMac);

          soma += bitsDiferentes(h1, h2) / (tamanhoSaida * 8);
        }
        final media = soma / _mensagensPorCombinacao;

        somaGlobal += media;
        combinacoes++;
        if (media < piorValor) {
          piorPosicao = pos;
          piorBit = bit;
          piorValor = media;
        }
      }
    }

    final mediaGlobal = somaGlobal / combinacoes;
    final vulneravel =
        mediaGlobal < _limiteMedia || piorValor < _limitePiorCaso;

    final Severidade severidade;
    if (!vulneravel) {
      severidade = Severidade.info;
    } else if (piorValor < 0.3) {
      severidade = Severidade.alta;
    } else {
      severidade = Severidade.media;
    }

    return ResultadoAtaque(
      nome(),
      vulneravel,
      severidade,
      'média global=${(mediaGlobal * 100).toStringAsFixed(1)}%; '
          'pior caso=${(piorValor * 100).toStringAsFixed(1)}% na '
          'entrada[posição=$piorPosicao, bit=$piorBit] sobre '
          '$tamanhoSaida bytes de MAC '
          '(limites: média>=${(_limiteMedia * 100).toStringAsFixed(0)}%, '
          'pior>=${(_limitePiorCaso * 100).toStringAsFixed(0)}%)',
      {
        'media_global': mediaGlobal,
        'pior': {
          'posicao': piorPosicao,
          'bit': piorBit,
          'valor': piorValor,
        },
      },
    );
  }
}

File: packages/julia/src/Seguranca/Ataques/AtaqueAvalancheChecksum.jl
"""
O AtaqueColisaoChecksum verifica se há colisões; esse verifica a difusão
interna: mudar 1 bit da MENSAGEM autenticada deveria mudar ~50% dos bits
dos 32 bytes de MAC, em QUALQUER posição. Um efeito avalanche fraco numa
posição específica significa que aquele byte quase não influencia o MAC -
uma pista forte de que a mistura tem um "ponto cego" estrutural.

Varre TODAS as posições/bits da entrada (não amostra aleatória) para
apontar exatamente onde está o ponto fraco.
"""
struct AtaqueAvalancheChecksum <: Ataque
    mensagens_por_combinacao::Int
    tamanho_entrada::Int
    limite_media::Float64
    limite_pior_caso::Float64
end

AtaqueAvalancheChecksum(;
    mensagens_por_combinacao::Integer = 10,
    tamanho_entrada::Integer = 32,
    limite_media::Real = 0.4,
    limite_pior_caso::Real = 0.4,
) = AtaqueAvalancheChecksum(
    Int(mensagens_por_combinacao),
    Int(tamanho_entrada),
    Float64(limite_media),
    Float64(limite_pior_caso),
)

nome(::AtaqueAvalancheChecksum) = "Efeito avalanche do checksum/MAC"

function executar(a::AtaqueAvalancheChecksum, alvo::AlvoCriptografico)
    chave_mac = repeat("M", 32)

    primeiro = checksum_bruto(alvo, "teste", chave_mac)
    tamanho_saida = length(primeiro)

    soma_global = 0.0
    combinacoes = 0
    pior_posicao = -1
    pior_bit = -1
    pior_valor = 1.0

    for pos in 0:(a.tamanho_entrada-1)
        for bit in 0:7
            soma = 0.0
            for _ in 1:(a.mensagens_por_combinacao)
                entrada = random_bytes(a.tamanho_entrada)
                alterada = copy(entrada)
                alterada[pos+1] = alterada[pos+1] ⊻ UInt8(1 << bit)

                h1 = checksum_bruto(alvo, entrada, chave_mac)
                h2 = checksum_bruto(alvo, alterada, chave_mac)

                soma += bits_diferentes(h1, h2) / (tamanho_saida * 8)
            end
            media = soma / a.mensagens_por_combinacao

            soma_global += media
            combinacoes += 1
            if media < pior_valor
                pior_posicao = pos
                pior_bit = bit
                pior_valor = media
            end
        end
    end

    media_global = soma_global / combinacoes
    vulneravel = media_global < a.limite_media || pior_valor < a.limite_pior_caso

    severidade = if !vulneravel
        "info"
    elseif pior_valor < 0.3
        "alta"
    else
        "media"
    end

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        severidade,
        "média global=$(fmt(media_global * 100, 1))%; pior caso=$(fmt(pior_valor * 100, 1))% na entrada[posição=$pior_posicao, bit=$pior_bit] sobre $tamanho_saida bytes de MAC (limites: média>=$(fmt(a.limite_media * 100, 0))%, pior>=$(fmt(a.limite_pior_caso * 100, 0))%)",
        Dict{String,Any}(
            "media_global" => media_global,
            "pior" => Dict{String,Any}(
                "posicao" => pior_posicao,
                "bit" => pior_bit,
                "valor" => pior_valor,
            ),
        ),
    )
end
